pkg/reconciliation/reconciliation.go

```diff
@@ -0,0 +1,53 @@
+package reconciliation
+
+import (
+	"context"
+	"time"
+
+	"github.com/k8ssandra/k8ssandra-operator/pkg/annotations"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/result"
+	errors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type k8sResource interface {
+	client.Object
+	annotations.Annotated
+	DeepCopy()
+	DeepCopyInto(out *client.Object)
+}
+
+func ReconcileObject[G k8sResource](ctx context.Context,
+	desiredObj k8sResource,
+	currentObj *k8sResource,
+	client client.Client,
+	requeueDelay time.Duration) result.ReconcileResult {
+	objectKey := types.NamespacedName{
+		Name:      desiredObj.GetName(),
+		Namespace: desiredObj.GetNamespace(),
+	}
+	annotations.AddHashAnnotation(desiredObj)
+	err := client.Get(ctx, objectKey, *currentObj)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if err := client.Create(ctx, desiredObj); err != nil {
+				return result.Error(err)
+			}
+			return result.RequeueSoon(requeueDelay)
+		} else {
+			return result.Error(err)
+		}
+	}
+
+	if !annotations.CompareHashAnnotations(*currentObj, desiredObj) {
+		resourceVersion := (*currentObj).GetResourceVersion()
+		desiredObj.DeepCopyInto(currentObj)
+		currentObj.SetResourceVersion(resourceVersion)
+		if err := client.Update(ctx, currentObj); err != nil {
+			return result.Error(err)
+		}
+		return result.RequeueSoon(requeueDelay)
+	}
+	return result.Done()
```

Comment:
This is more of a question than a change request. Would it be better to return `result.Continue()` here? I can see both ways. I guess it just depends on who should own the logic of deciding when a reconcile is done.

Reply:
I'm certainly open to that. I guess the reconciliation of this particular object is complete, so `Done()` made some sense here. We should comment the reconcile result types so developers can have some sense of what they're intended to mean.

Reply:
Yes, for what you are doing here, I think `Done()` makes the most sense.